tests.py

```diff
@@ -1,81 +1,241 @@
+    Assumptions: The strings in this project are necessarily in the same length.
+    :param string1:
+    :param string2:
+    :return: the distance
+    """
+    count_diffs = 0
+    for a, b in zip(string1, string2):
+        if a != b:
+            count_diffs += 1
+    return count_diffs
+
+
+# CLASSES
+
+class TestClassCheckAPI(unittest.TestCase):
+    """
+    This test class was created in order to check the API that was given in this exercise.
+    """
+
+    def setUp(self):
+        """
+        the testing framework executes the test setup method first,
+        before any test method in the class. It includes all the required data for each test.
+        """
+        self.url = "https://hangman-api.herokuapp.com/hangman"
+        self.response = requests.post(self.url, data='text=great').json()
+        self.game_token = self.response['token']
+        self.hangman_string = self.response['hangman']
+
+    def test_types(self):
+        """
+        Checks if the API sends valid & expected types: JSON, DICT etc.
+        :return:
+        """
+        response = requests.post(self.url, data='text=great')
+        assert str(type(response)) == "<class 'requests.models.Response'>"
+        response = response.json()
+        assert str(type(response)) == "<class 'dict'>"
+
+    def test_valid_login(self):
+        response = requests.post(self.url, data='text=great')
+        assert response.status_code == 200
+        keys = set(response.json().keys())
+        assert keys == {'hangman', 'token'}
+
+    def test_valid_get_hint(self):
+        response = requests.get(self.url + '/hint', data={'token': self.game_token})
+        assert response.status_code == 200
+        keys = set(response.json().keys())
+        assert keys == {'hint', 'token'}
+
+    def test_valid_guess(self):
+        response = requests.put(self.url, data={'token': self.game_token, 'letter': 'k'})
+        assert response.status_code == 200
+        keys = set(response.json().keys())
+        assert keys == {'hangman', 'token', 'correct'}
+
+    def test_repeating_guess(self):
+        response = requests.put(self.url, data={'token': self.game_token, 'letter': 'l'})
+        game_token = response.json()['token']
+        response = requests.put(self.url, data={'token': game_token, 'letter': 'l'})
+        assert response.status_code == 304
+
+    def test_invalid_guess(self):
+        response = requests.put(self.url, data={'token': self.game_token, 'letter': 'dl'})
+        assert response.status_code != 200
+
+    def test_valid_get_solution(self):
+        response = requests.get(self.url, data={'token': self.game_token})
+        assert response.status_code == 200
+        keys = set(response.json().keys())
+        assert keys == {'solution', 'token'}
+
+    def tearDown(self):
+        self.test_data = []
+
+
+class TestClassGame(unittest.TestCase):
+
+    def setUp(self):
```

Comment:
This is not so important in that situation, and yet I want to point out a pattern that I think should be avoided next time: we create a lot of objects in the setup, but use only few in each test. We should avoid wasting memory and computation in general. Of course it is extremely cheap in this case, but it might look to the reader as if you need all of these objects in every test. Let's think if we can find a better solution here. What do you say?